### PR TITLE
[SYCL][SPIR-V] Make clang lookup for SPIR-V builtins

### PR DESCRIPTION
**Pull request by Naghasan**
This patch (2 commits) add the initial infrastructure to enable lookup of SPIR-V builtins by clang following a similar model as done for OpenCL.

The first commit is to make the OpenCL lookup agnostic to the programming model used. The second enables the path for them and add a dummy instructions as example. Other builtins will follow after this.

Down the line, this will enforce a stable way to express SPIR-V builtins and make them closer to how the translator mangles them. This will help ensuring builtin for CUDA does not break easily. This will also support any changes suggested by the SPIRV-LLVM people on how to represent builtins.

Signed-off-by: Victor Lomuller <victor@codeplay.com>

### CODE REVIEW

File: clang/test/SemaSYCL/spirv-builtin-lookup.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -fdeclare-spirv-builtins -fsyntax-only -verify -std=c++11 %s
```
Comment (bader) [inline clang/test/SemaSYCL/spirv-builtin-lookup.cpp]:
```suggestion
// RUN: %clang_cc1 -fdeclare-spirv-builtins -fsyntax-only -verify %s
```

File: clang/test/SemaSYCL/spirv-builtin-lookup.cpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// expected-no-diagnostics
+
+// Verify that __spirv_ocl_acos is recognized as a builtin
+
+float acos(float val) {
+  return __spirv_ocl_acos(val);
+}
+
+double acos(double val) {
+  return __spirv_ocl_acos(val);
```
Comment (bader) [inline clang/test/SemaSYCL/spirv-builtin-lookup.cpp]: Could you also add a CodeGen test, please?

File: clang/test/SemaSYCL/spirv-builtin-lookup-invalid.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -fdeclare-spirv-builtins -fsyntax-only -verify -std=c++11 %s
```
Comment (bader) [inline clang/test/SemaSYCL/spirv-builtin-lookup-invalid.cpp]:
```suggestion
// RUN: %clang_cc1 -fdeclare-spirv-builtins -fsyntax-only -verify %s
```

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp [hunk 4]
```diff
@@ -1,4 +1,4 @@
-//===- ClangOpenCLBuiltinEmitter.cpp - Generate Clang OpenCL Builtin handling
+//===- ClangProgModelBuiltinEmitter.cpp - Generate Clang Builtin handling
```
Comment (bader) [inline clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: The file should be renamed accordingly.

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp [hunk 5]
```diff
@@ -83,24 +83,32 @@ struct BuiltinTableEntries {
...
   // Record instances. They are coming from the associated TableGen file.
   RecordKeeper &Records;
 
   // The output file.
   raw_ostream &OS;
 
+  // Family for which the builtin are for.
+  llvm::StringRef Family;
+
+  // Family for which the builtin are for.
```
Comment (bader) [inline clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: Looks like a copy of the comment at line 106.
```suggestion
  // Class for which the builtin are for.
```
?

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp [hunk 6]
```diff
@@ -448,32 +473,32 @@ void BuiltinNameEmitter::EmitExtensionTable() {
...
 }
 
 void BuiltinNameEmitter::EmitSignatureTable() {
   // Store a type (e.g. int, float, int2, ...). The type is stored as an index
-  // of a struct OpenCLType table. Multiple entries following each other form a
-  // signature.
-  OS << "static const unsigned short SignatureTable[] = {\n";
+  // of a struct ProgModelTypeStruct table. Multiple entries following each
+  // other form a signature.
+  OS << "const unsigned short " << ClassName << "::SignatureTable[] = {\n";
```
Comment (bader) [inline clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: Why `static` was removed?

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp (hunk 4 above)
Comment (Naghasan) [reply clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: I agree, I'll do that once approved if you don't mind. I was trying to make sure the changes in the file remains clearly visible (renaming the file may affect the way github display it).

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp (hunk 6 above)
Comment (Naghasan) [reply clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: upstream emit the lookup data as simple static data (as only OpenCL is of any interest).

`SignatureTable` (others) are now static fields of a class and this the implementation. The declaration is here https://github.com/intel/llvm/pull/1345/files/b07b2a86fa88a176c8c3115e35c0742fd4d805f3#diff-4326db80135f293472deaabc7779589fR365

File: clang/test/SemaSYCL/spirv-builtin-lookup.cpp (hunk 2 above)
Comment (Naghasan) [reply clang/test/SemaSYCL/spirv-builtin-lookup.cpp]: Yes, done here https://github.com/intel/llvm/pull/1345/commits/d4bec11b8dbe0610a2066ad6c70460006f6b9bf9

File: clang/test/SemaSYCL/spirv-builtin-lookup-invalid.cpp (hunk 3 above)
Comment (Naghasan) [reply clang/test/SemaSYCL/spirv-builtin-lookup-invalid.cpp]: Changes applied here https://github.com/intel/llvm/pull/1345/commits/9c820a05ae8ba16bf601fc71536fd5edc91ffe3b

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp (hunk 4 above)
Comment (bader) [reply clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: Here GitHub was able to handle both renaming + content changes: https://github.com/intel/llvm/pull/1306/files#diff-e4c028e8ba11e842d27a835ad03e47be.
Can we give it a try?
You can do it in a separate commit and revert if it doesn't work as expected.

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp (hunk 6 above)
Comment (bader) [reply clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: I see. Thanks for clarification.

File: clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp (hunk 4 above)
Comment (Naghasan) [reply clang/utils/TableGen/ClangOpenCLBuiltinEmitter.cpp]: TIL it actually works in github :)

**Comment by bader** [comment]
@Naghasan, could you update the Windows mangling in CodeGen test, please?

**Comment by Naghasan** [comment]
@bader I tests updated

**Pull request merged**